File: task7/src/test/java/project/pages/MainPage.java
package project.pages;

import aquality.selenium.elements.interfaces.IButton;
import aquality.selenium.elements.interfaces.ILabel;
import aquality.selenium.forms.Form;
import aquality.selenium.waitings.ConditionalWait;
import framework.utils.RandomNumber;
import org.openqa.selenium.By;
import org.openqa.selenium.support.ui.ExpectedConditions;
import project.models.Field;

public class MainPage extends Form {

    public MainPage() {
        super(By.xpath("//div[@class='notifications-wrap']"), "Main page");
    }

    private final IButton btnRandomOpponent = getElementFactory().getButton(By.xpath("//a[@class='battlefield-start-choose_rival-variant-link']"), "Random opponent");
    private final IButton btnRandomPlaceShip = getElementFactory().getButton(By.xpath("//ul[@class='placeships']//span"), "Random place ship");
    private final IButton btnStartGame = getElementFactory().getButton(By.xpath("//div[@class='battlefield-start-button']"), "Button start game");
    private final ILabel labelGame = getElementFactory().getLabel(By.xpath(new StringBuilder().append("//div[contains(@class, 'rival-leave') ")
            .append("or contains(@class,'game-over-win') ")
            .append("or contains(@class,'game-over-lose') ")
            .append("or contains(@class,'server-error') ")
            .append("or contains(@class,'game-error')]").toString()), "Label notification");
    private static final String emptyCellAttribute = "battlefield-cell__empty";
    private static final String hitCellAttribute = "battlefield-cell__hit";
    private static final String attributeField = "class";

    private IButton getBtnField(Field field) {
        return getElementFactory().getButton(By.xpath(String.format("//div[@class='battlefield battlefield__rival']//tr[%s]//td[%s]", field.getRow(), field.getCol())), "field " + field.getRow() + " " + field.getCol());
    }

    public void clickRandomPlaceShipRandomTimes(int bound) {
        for (int i = 0; i < RandomNumber.randomInRange(bound); i++) {
            btnRandomPlaceShip.click();
        }
    }

    public void clickRandomOpponent() {
        btnRandomOpponent.click();
    }

    public void clickStartGame() {
        btnStartGame.click();
    }

    public void clickField(Field field) {
        if (getBtnField(field).getAttribute(attributeField).contains(emptyCellAttribute)) {
            getBtnField(field).click();
        }
    }

    public boolean isInjureShip(Field field) {
        return getBtnField(field).getAttribute(attributeField).contains(hitCellAttribute);
    }

    public String getTextFromLabelGame() {
        return labelGame.getText();
    }

    public String getAttributeFromLabelGame() {
        ConditionalWait.waitFor(ExpectedConditions.visibilityOf(labelGame.getElement()), "Wait Label");
        return labelGame.getAttribute(attributeField);
    }

    public boolean isInvisibleLabel() {
        return !labelGame.state().isDisplayed();
    }
}

File: ButtleShip/src/test/kotlin/pages/MainPage.kt
package pages

import framework.elements.Button
import framework.elements.Label
import framework.pages.BasePage
import framework.utils.randomInRange
import framework.waitings.waitFor
import models.Field
import org.openqa.selenium.By
import org.openqa.selenium.support.ui.ExpectedConditions

class MainPage : BasePage(By.xpath("//div[@class='notifications-wrap']"), "Main page") {

    private val btnRandomOpponent = Button(
        By.xpath("//a[@class='battlefield-start-choose_rival-variant-link']"),
        "Random opponent"
    )
    private val btnRandomPlaceShip =
        Button(By.xpath("//ul[@class='placeships']//span"), "Random place ship")
    private val btnStartGame =
        Button(By.xpath("//div[@class='battlefield-start-button']"), "Button start game")
    private val labelGame = Label(
        By.xpath(
            StringBuilder().append("//div[contains(@class, 'game-error') ")
                .append("or contains(@class,'game-over-win') ")
                .append("or contains(@class,'game-over-lose')]").toString()
        ), "Label notification"
    )
    private val emptyCellAttribute = "battlefield-cell__empty"
    private val hitCellAttribute = "battlefield-cell__hit"
    private val attributeField = "class"

    private fun getBtnField(field: Field): Button {
        return Button(
            By.xpath("//div[@class='battlefield battlefield__rival']//tr[${field.row}]//td[${field.col}]"),
            "field ${field.row} ${field.col}"
        )
    }

    fun clickRandomPlaceShipRandomTimes(bound: Int) {
        for (i in 0..randomInRange(bound)) {
            btnRandomPlaceShip.click()
        }
    }

    fun clickRandomOpponent() {
        btnRandomOpponent.click()
    }

    fun clickStartGame() {
        btnStartGame.click()
    }

    fun clickField(field: Field) {
        if (getBtnField(field).getAttribute(attributeField)!!.contains(emptyCellAttribute)) {
            getBtnField(field).click()
        }
    }

    fun isInjureShip(field: Field): Boolean {
        return getBtnField(field).getAttribute(attributeField)!!.contains(hitCellAttribute)
    }

    fun getTextFromLabelGame(): String? {
        return labelGame.getText()
    }

    fun getAttributeFromLabelGame(): String? {
        waitFor(
            ExpectedConditions.visibilityOf(labelGame.getElement()),
            "Wait Label"
        )
        return labelGame.getAttribute(attributeField)
    }

    fun isVisibleLabel(): Boolean {
        return labelGame.isDisplayed()
    }
}